packages/flutter/lib/src/widgets/tooltip.dart

```diff
@@ -0,0 +1,819 @@
+  /// visibility.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  /// Determines the visibility of [Tooltip] widgets that inherit from this widget.
+  final bool visible;
+
+  /// The [visible] of the closest instance of this class that encloses the
+  /// given context. Defaults to `true` if none are found.
+  static bool of(BuildContext context) {
+    final _TooltipVisibilityScope? visibility = context
+        .dependOnInheritedWidgetOfExactType<_TooltipVisibilityScope>();
+    return visibility?.visible ?? true;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return _TooltipVisibilityScope(visible: visible, child: child);
+  }
+}
+
+/// The method of interaction that will trigger a tooltip.
+/// Used in [RawTooltip.triggerMode].
+///
+/// On desktop, a tooltip will be shown as soon as a pointer hovers over
+/// the widget, regardless of the value of [RawTooltip.triggerMode].
+///
+/// See also:
+///
+///   * [RawTooltip.waitDuration], which defines the length of time that
+///     a pointer must hover over a tooltip's widget before the tooltip
+///     will be shown.
+enum TooltipTriggerMode {
+  /// Tooltip will only be shown by calling `ensureTooltipVisible`.
+  manual,
+
+  /// Tooltip will be shown after a long press.
+  ///
+  /// See also:
+  ///
+  ///   * [GestureDetector.onLongPress], the event that is used for trigger.
+  ///   * [Feedback.forLongPress], the feedback method called when feedback is enabled.
+  longPress,
+
+  /// Tooltip will be shown after a single tap.
+  ///
+  /// See also:
+  ///
+  ///   * [GestureDetector.onTap], the event that is used for trigger.
+  ///   * [Feedback.forTap], the feedback method called when feedback is enabled.
+  tap,
+}
+
+/// Signature for when a tooltip is triggered.
+typedef TooltipTriggeredCallback = void Function();
+
+/// A special [MouseRegion] that when nested, only the first [_ExclusiveMouseRegion]
+/// to be hit in hit-testing order will be added to the BoxHitTestResult (i.e.,
+/// child over parent, last sibling over first sibling).
+///
+/// The [onEnter] method will be called when a mouse pointer enters this
+/// [MouseRegion], and there is no other [_ExclusiveMouseRegion]s obstructing
+/// this [_ExclusiveMouseRegion] from receiving the events. This includes the
+/// case where the mouse cursor stays within the paint bounds of an outer
+/// [_ExclusiveMouseRegion], but moves outside of the bounds of the inner
+/// [_ExclusiveMouseRegion] that was initially blocking the outer widget.
+///
+/// Likewise, [onExit] is called when the a mouse pointer moves out of the paint
+/// bounds of this widget, or moves into another [_ExclusiveMouseRegion] that
+/// overlaps this widget in hit-testing order.
+///
+/// This widget doesn't affect [MouseRegion]s that aren't [_ExclusiveMouseRegion]s,
+/// or other [HitTestTarget]s in the tree.
+class _ExclusiveMouseRegion extends MouseRegion {
```

Comment:
Like you mentioned in the draft design doc, `RawTooltip` is visually unopinionated, but behaviorally opinionated. If a custom design system needs a different behavior, they might need to fork the logic in `RawTooltip`. 

Do you think `_ExclusiveMouseRegion`, `_TooltipOverlay`, and `_TooltipPositionDelegate` make sense as public APIs? Assuming these can be standalone APIs, that would make it easier to fork `RawTooltip` to customize its behavior.

Reply:
To me a tooltip consists of:

 * The UI that shows the tooltip when you hover it (`child`).
 * The tooltip itself (tooltipBuilder). Could be split between the frame and the contents but that's probably for a design language to do.
 * The positioning logic (_TooltipOverlay, which includes _TooltipPositionDelegate).
 * The logic for how to listen to user events and decide when to show the tooltip (_ExclusiveMouseRegion plus Listener).
 * Semantics.

I think a user could reasonably want to customize any of those things besides Semantics. It would be cool to write a RawTooltip that allowed you to customize all of that, but that's a tall order, and it's not easy to get it right. You could always delay any/all of that to another PR, or even wait for feedback from users.